### ISSUE TITLE
base_linux: segmention fault in ld.lib.so

### ISSUE THREAD

**Issue by alex-ab** [genodelabs/genode#3797]
Sporadically, I get segfaults in ld.lib.so (genode master 4450b37ff5b22e99133863a6f56d5b715c7f3aaa, scrcp_rgb888)
```
[init -> scrcpy] lwIP Nic interface up address=10.0.3.2 netmask=255.255.255.0 gateway=10.0.3.1
[init -> scrcpy] create thread stream
[init -> scrcpy] create pthread.1 -> cpu 0
[init -> scrcpy] create thread controller
[init -> scrcpy] create pthread.2 -> cpu 0
[init -> scrcpy] create pthread.3 -> cpu 0
[init -> scrcpy] create thread receiver
[init -> scrcpy] create pthread.4 -> cpu 2
Kernel: Segmentation fault (signum=11), see Linux kernel log for detail
```

```
[ 7593.836907] audit: type=1326 audit(1593597066.198:50): auid=1000 uid=1000 gid=1000 ses=3 pid=601 comm="ld.lib.so" exe="build/linux/var/run/scrcpy/genode/ld.lib.so" sig=31 arch=c000003e syscall=15 compat=0 ip=0x500aba67 code=0x80000000
```

```
00000000500aba5e <lx_restore_rt>:
lx_restore_rt():
repos/base-linux/src/lib/syscall/spec/x86_64/lx_restore_rt.S:15
    500aba5e:   48 c7 c0 0f 00 00 00    mov    $0xf,%rax
repos/base-linux/src/lib/syscall/spec/x86_64/lx_restore_rt.S:16
    500aba65:   0f 05                   syscall 

00000000500aba67 <lx_syscall>:
lx_syscall():
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:20
    500aba67:   48 89 f8                mov    %rdi,%rax
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:21
    500aba6a:   48 89 f7                mov    %rsi,%rdi
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:22
    500aba6d:   48 89 d6                mov    %rdx,%rsi
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:23
    500aba70:   48 89 ca                mov    %rcx,%rdx
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:24
    500aba73:   4d 89 c2                mov    %r8,%r10
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:25
    500aba76:   4d 89 c8                mov    %r9,%r8
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:26
    500aba79:   4c 8b 4c 24 08          mov    0x8(%rsp),%r9
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:27
    500aba7e:   0f 05                   syscall 

00000000500aba80 <pseudo_end>:
pseudo_end():
repos/base-linux/src/lib/syscall/spec/x86_64/lx_syscall.S:29
    500aba80:   c3                      retq   
```

**Comment by chelmuth** [comment]
Unfortunately, the kernel-log line tells us that `syscall=15` which is  `rt_sigreturn` failed. Is there more in the log before this message?

**Comment by alex-ab** [comment]
Not really:

```
[ 5432.784797] CPU2: Package temperature/speed normal
[ 5986.446253] [drm] Reducing the compressed framebuffer size. This may lead to less power savings than a non-reduced-size. Try to increase stolen memory size if available in BIOS.
[ 6601.833842] [drm] Reducing the compressed framebuffer size. This may lead to less power savings than a non-reduced-size. Try to increase stolen memory size if available in BIOS.
[ 6840.116627] IPv6: ADDRCONF(NETDEV_CHANGE): tap0: link becomes ready
[ 7593.836907] audit: type=1326 audit(1593597066.198:50): auid=1000 uid=1000 gid=1000 ses=3 pid=601 comm="ld.lib.so" exe="/home/alex/genode/genode.devel/build/linux/var/run/scrcpy/genode/ld.lib.so" sig=31 arch=c
000003e syscall=15 compat=0 ip=0x500aba67 code=0x80000000
```

**Comment by alex-ab** [comment]
> Unfortunately, the kernel-log line tells us that `syscall=15` which is `rt_sigreturn` failed. Is there more in the log before this message?

I updated the objdump snippet. As @ssumpf pointed out, the rt_sigreturn code is before the fault ip. Is after the syscall something like ret* missing ?

**Comment by chelmuth** [comment]
As I understand it the `rt_sigreturn` is a return from signal handler to the normal execution and should never return to the calling IP, ie. `lx_restore_rt` is `noreturn`.

**Comment by chelmuth** [comment]
I found https://bugs.launchpad.net/ubuntu-z-systems/+bug/1686618 which shares the cryptic log message with this issue (`sig=31...`). @throwException Is this an indicator for a missing syscall permission in seccomp?

**Comment by chelmuth** [comment]
@alex-ab please try to copy https://github.com/genodelabs/genode/blob/4450b37ff5b22e99133863a6f56d5b715c7f3aaa/tool/seccomp/seccomp_bpf_compiler.h#L183-L184 to the x86 path and retry.

**Comment by throwException** [comment]
> I found https://bugs.launchpad.net/ubuntu-z-systems/+bug/1686618 which shares the cryptic log message with this issue (`sig=31...`). @throwException Is this an indicator for a missing syscall permission in seccomp?
There should be no segfault when seccomp terminates a process as SIGSYS is sent to all threads at once.

> @alex-ab please try to copy
> https://github.com/genodelabs/genode/blob/4450b37ff5b22e99133863a6f56d5b715c7f3aaa/tool/seccomp/seccomp_bpf_compiler.h#L183-L184
Beware to update the binary BPF filter using `repos/base-linux/src/lib/seccomp/update.sh`.

> [ 7593.836907] audit: type=1326 audit(1593597066.198:50): auid=1000 uid=1000 gid=1000 ses=3 pid=601 comm="ld.lib.so" exe="build/linux/var/run/scrcpy/genode/ld.lib.so" sig=31 arch=c000003e syscall=15 compat=0 ip=0x500aba67 code=0x80000000
Have you enabled audit in `seccomp_bpf_compiler.h`? Otherwise I wouldn't expect a message in `dmesg` as this feature is not enable by default as it's not available on all targets.

As a side note: To debug seccomp, `strace` is very helpful, especially if all known good syscalls are filtered out.

**Comment by chelmuth** [comment]
> Have you enabled audit in `seccomp_bpf_compiler.h`? Otherwise I wouldn't expect a message in `dmesg` as this feature is not enable by default as it's not available on all targets.

It seems my Ubuntu 18.04 system is that verbose by default.
```
> cat /proc/sys/kernel/seccomp/actions_logged
kill_process kill_thread trap errno trace log
```

> As a side note: To debug seccomp, `strace` is very helpful, especially if all known good syscalls are filtered out.

Are you referring to `strace -p <pid>`? Do you have an example command line with all filters specified?

**Comment by throwException** [comment]
> Are you referring to `strace -p <pid>`? Do you have an example command line with all filters specified?

Yes, I change `tool/run/power_on/linux` with the following line to debug:
`eval spawn strace -e !sendmsg,recvmsg,close,epoll_wait,epoll_ctl,epoll_create,socketpair,mmap,dup,dup2,futex,nanosleep,gettimeofday,fstat,getpid,munmap,rt_sigaction -ff -o /tmp/genode ./core`

**Comment by chelmuth** [comment]
Thanks, this looks quite useful.

**Comment by alex-ab** [comment]
Addressed by #3799
